# **01 Churn prediction project**
- Dataset: https://www.kaggle.com/blastchar/telco-customer-churn
- https://raw.githubusercontent.com/alexeygrigorev/mlbookcamp-code/master/chapter-03-churn-prediction/WA_Fn-UseC_-Telco-Customer-Churn.csv

We'll use logistic regression to predict churn

# 02 Data Preparation

In [60]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt

In [61]:
df = pd.read_csv('WA_Fn-UseC_-Telco-Customer-Churn.csv')
df.head()

customerID  gender  SeniorCitizen Partner Dependents  tenure PhoneService  \
0  7590-VHVEG  Female              0     Yes         No       1           No   
1  5575-GNVDE    Male              0      No         No      34          Yes   
2  3668-QPYBK    Male              0      No         No       2          Yes   
3  7795-CFOCW    Male              0      No         No      45           No   
4  9237-HQITU  Female              0      No         No       2          Yes   

      MultipleLines InternetService OnlineSecurity  ... DeviceProtection  \
0  No phone service             DSL             No  ...               No   
1                No             DSL            Yes  ...              Yes   
2                No             DSL            Yes  ...               No   
3  No phone service             DSL            Yes  ...              Yes   
4                No     Fiber optic             No  ...               No   

  TechSupport StreamingTV StreamingMovies        Contract PaperlessBilling  \
0          No          No              No  Month-to-month              Yes   
1          No          No              No        One year               No   
2          No          No              No  Month-to-month              Yes   
3         Yes          No              No        One year               No   
4          No          No              No  Month-to-month              Yes   

               PaymentMethod MonthlyCharges  TotalCharges Churn  
0           Electronic check          29.85         29.85    No  
1               Mailed check          56.95        1889.5    No  
2               Mailed check          53.85        108.15   Yes  
3  Bank transfer (automatic)          42.30       1840.75    No  
4           Electronic check          70.70        151.65   Yes  

[5 rows x 21 columns]

In [62]:
df.columns

Index(['customerID', 'gender', 'SeniorCitizen', 'Partner', 'Dependents',
       'tenure', 'PhoneService', 'MultipleLines', 'InternetService',
       'OnlineSecurity', 'OnlineBackup', 'DeviceProtection', 'TechSupport',
       'StreamingTV', 'StreamingMovies', 'Contract', 'PaperlessBilling',
       'PaymentMethod', 'MonthlyCharges', 'TotalCharges', 'Churn'],
      dtype='object')

In [63]:
df.columns = df.columns.str.lower().str.replace(' ', '_')
catagorical_columns = list(df.dtypes[df.dtypes == 'object'].index)
for c in catagorical_columns:
    df[c] = df[c].str.lower().str.replace(' ', '_')

In [64]:
df.head()

customerid  gender  seniorcitizen partner dependents  tenure phoneservice  \
0  7590-vhveg  female              0     yes         no       1           no   
1  5575-gnvde    male              0      no         no      34          yes   
2  3668-qpybk    male              0      no         no       2          yes   
3  7795-cfocw    male              0      no         no      45           no   
4  9237-hqitu  female              0      no         no       2          yes   

      multiplelines internetservice onlinesecurity  ... deviceprotection  \
0  no_phone_service             dsl             no  ...               no   
1                no             dsl            yes  ...              yes   
2                no             dsl            yes  ...               no   
3  no_phone_service             dsl            yes  ...              yes   
4                no     fiber_optic             no  ...               no   

  techsupport streamingtv streamingmovies        contract paperlessbilling  \
0          no          no              no  month-to-month              yes   
1          no          no              no        one_year               no   
2          no          no              no  month-to-month              yes   
3         yes          no              no        one_year               no   
4          no          no              no  month-to-month              yes   

               paymentmethod monthlycharges  totalcharges churn  
0           electronic_check          29.85         29.85    no  
1               mailed_check          56.95        1889.5    no  
2               mailed_check          53.85        108.15   yes  
3  bank_transfer_(automatic)          42.30       1840.75    no  
4           electronic_check          70.70        151.65   yes  

[5 rows x 21 columns]

In [65]:
df.dtypes

customerid           object
gender               object
seniorcitizen         int64
partner              object
dependents           object
tenure                int64
phoneservice         object
multiplelines        object
internetservice      object
onlinesecurity       object
onlinebackup         object
deviceprotection     object
techsupport          object
streamingtv          object
streamingmovies      object
contract             object
paperlessbilling     object
paymentmethod        object
monthlycharges      float64
totalcharges         object
churn                object
dtype: object

In [66]:
tc = pd.to_numeric(df.totalcharges, errors='coerce')

In [67]:
df.totalcharges = pd.to_numeric(df.totalcharges, errors='coerce')
df.totalcharges = df.totalcharges.fillna(0)

In [68]:
df[tc.isnull()][['customerid', 'totalcharges']]

customerid  totalcharges
488   4472-lvygi           0.0
753   3115-czmzd           0.0
936   5709-lvoeq           0.0
1082  4367-nuyao           0.0
1340  1371-dwpaz           0.0
3331  7644-omvmy           0.0
3826  3213-vvolg           0.0
4380  2520-sgtta           0.0
5218  2923-arzlg           0.0
6670  4075-wkniu           0.0
6754  2775-sefee           0.0

In [69]:
df.churn = (df.churn == 'yes').astype(int)

# 03 Setting up the validation framework
- Perform the train/validation/test split with Scikit-Learn

In [70]:
from sklearn.model_selection import train_test_split

In [71]:
df_full_train, df_test = train_test_split(df, test_size=0.2, random_state=1)

In [72]:
len(df_full_train), len(df_test)

(5634, 1409)

In [73]:
# here to take 20% of validation data from df_full_train we have to set test_size as 0.25 i.e. 25% of full 80% dataset
df_train, df_val = train_test_split(df_full_train, test_size=0.25, random_state=1)

In [74]:
len(df_train), len(df_val), len(df_test)

(4225, 1409, 1409)

In [75]:
df_train = df_train.reset_index(drop=True)
df_val = df_val.reset_index(drop=True)
df_test = df_test.reset_index(drop=True)

In [76]:
df_train.head()

customerid  gender  seniorcitizen partner dependents  tenure phoneservice  \
0  8015-ihcgw  female              0     yes        yes      72          yes   
1  1960-uycnn    male              0      no         no      10          yes   
2  9250-wypll  female              0      no         no       5          yes   
3  6786-obwqr  female              0     yes        yes       5          yes   
4  1328-euzhc  female              0     yes         no      18          yes   

  multiplelines internetservice       onlinesecurity  ...  \
0           yes     fiber_optic                  yes  ...   
1           yes     fiber_optic                   no  ...   
2           yes     fiber_optic                   no  ...   
3            no     fiber_optic                   no  ...   
4            no              no  no_internet_service  ...   

      deviceprotection          techsupport          streamingtv  \
0                  yes                  yes                  yes   
1                  yes                   no                   no   
2                   no                   no                   no   
3                   no                   no                  yes   
4  no_internet_service  no_internet_service  no_internet_service   

       streamingmovies        contract paperlessbilling     paymentmethod  \
0                  yes        two_year              yes  electronic_check   
1                  yes  month-to-month              yes  electronic_check   
2                   no  month-to-month               no  electronic_check   
3                   no  month-to-month              yes  electronic_check   
4  no_internet_service        two_year               no      mailed_check   

  monthlycharges  totalcharges  churn  
0         115.50       8425.15      0  
1          95.25       1021.55      0  
2          75.55        413.65      1  
3          80.85        356.10      0  
4          20.10        370.50      0  

[5 rows x 21 columns]

In [77]:
# write churn values to y as we are going to predict churn
y_train = df_train.churn.values
y_val = df_val.churn.values
y_test = df_test.churn.values

In [78]:
del df_train['churn']
del df_val['churn']
del df_test['churn']

# 04 EDA
- Check missing values
- Look at the target variable (churn)
- Look at the numerical and catagorical variables

In [82]:
df_full_train = df_full_train.reset_index(drop=True)
df_full_train.head()

customerid  gender  seniorcitizen partner dependents  tenure phoneservice  \
0  5442-pptjy    male              0     yes        yes      12          yes   
1  6261-rcvns  female              0      no         no      42          yes   
2  2176-osjuv    male              0     yes         no      71          yes   
3  6161-erdgd    male              0     yes        yes      71          yes   
4  2364-ufrom    male              0      no         no      30          yes   

  multiplelines internetservice       onlinesecurity  ...  \
0            no              no  no_internet_service  ...   
1            no             dsl                  yes  ...   
2           yes             dsl                  yes  ...   
3           yes             dsl                  yes  ...   
4            no             dsl                  yes  ...   

      deviceprotection          techsupport          streamingtv  \
0  no_internet_service  no_internet_service  no_internet_service   
1                  yes                  yes                   no   
2                   no                  yes                   no   
3                  yes                  yes                  yes   
4                   no                  yes                  yes   

       streamingmovies  contract paperlessbilling              paymentmethod  \
0  no_internet_service  two_year               no               mailed_check   
1                  yes  one_year               no    credit_card_(automatic)   
2                   no  two_year               no  bank_transfer_(automatic)   
3                  yes  one_year               no           electronic_check   
4                   no  one_year               no           electronic_check   

  monthlycharges  totalcharges  churn  
0          19.70        258.35      0  
1          73.90       3160.55      1  
2          65.15       4681.75      0  
3          85.45       6300.85      0  
4          70.40       2044.75      0  

[5 rows x 21 columns]

In [83]:
df_full_train.isnull().sum()

customerid          0
gender              0
seniorcitizen       0
partner             0
dependents          0
tenure              0
phoneservice        0
multiplelines       0
internetservice     0
onlinesecurity      0
onlinebackup        0
deviceprotection    0
techsupport         0
streamingtv         0
streamingmovies     0
contract            0
paperlessbilling    0
paymentmethod       0
monthlycharges      0
totalcharges        0
churn               0
dtype: int64

In [86]:
df_full_train.churn.value_counts(normalize=True) # normalize is to show data in form of % values

churn
0    0.730032
1    0.269968
Name: proportion, dtype: float64

0.269968 -> this is called the churn rate which we can read as 26%

In [89]:
# mean value is same as churn % as mean is sum of ones and not zeros and hence
df_full_train.churn.mean()

np.float64(0.26996805111821087)

In [90]:
global_churn_rate = df_full_train.churn.mean()
round(global_churn_rate, 2)

np.float64(0.27)

will now look at other variables i.e. numerical variables and catagorical variables

In [97]:
df_full_train.dtypes

customerid           object
gender               object
seniorcitizen         int64
partner              object
dependents           object
tenure                int64
phoneservice         object
multiplelines        object
internetservice      object
onlinesecurity       object
onlinebackup         object
deviceprotection     object
techsupport          object
streamingtv          object
streamingmovies      object
contract             object
paperlessbilling     object
paymentmethod        object
monthlycharges      float64
totalcharges        float64
churn                 int64
dtype: object

In [99]:
numerical = ['tenure', 'monthlycharges', 'totalcharges']
catagorical = ['gender', 'seniorcitizen', 'partner', 'dependents',
       'phoneservice', 'multiplelines', 'internetservice',
       'onlinesecurity', 'onlinebackup', 'deviceprotection', 'techsupport',
       'streamingtv', 'streamingmovies', 'contract', 'paperlessbilling',
       'paymentmethod', 'churn']

In [101]:
df_full_train[catagorical].nunique()

gender              2
seniorcitizen       2
partner             2
dependents          2
phoneservice        2
multiplelines       3
internetservice     3
onlinesecurity      3
onlinebackup        3
deviceprotection    3
techsupport         3
streamingtv         3
streamingmovies     3
contract            3
paperlessbilling    2
paymentmethod       4
churn               2
dtype: int64

# 05 Feature importance: Churn rate and risk ratio
Feature importance analysis (part of EDA) - identifying which features affect our target variable
- Churn rate
- Risk ratio
- Mutual information - later

**Churn rate**

In [103]:
churn_female = df_full_train[df_full_train.gender == 'female'].churn.mean()
churn_female

np.float64(0.27682403433476394)

In [104]:
churn_male = df_full_train[df_full_train.gender == 'male'].churn.mean()
churn_male

np.float64(0.2632135306553911)

In [106]:
global_churn = df_full_train.churn.mean()
global_churn

np.float64(0.26996805111821087)

In [107]:
df_full_train.partner.value_counts()

partner
no     2932
yes    2702
Name: count, dtype: int64

In [109]:
churn_partner = df_full_train[df_full_train.partner == 'yes'].churn.mean()
churn_partner

np.float64(0.20503330866025166)

In [114]:
global_churn - churn_partner # value > 0 means chances of churn is less for this group

np.float64(0.06493474245795922)

In [115]:
churn_no_partner = df_full_train[df_full_train.partner == 'no'].churn.mean()
churn_no_partner

np.float64(0.3298090040927694)

In [116]:
global_churn - churn_no_partner # value < 0 means chances of churn is high for this group

np.float64(-0.05984095297455855)

**Risk Ratio**

In [117]:
churn_no_partner / global_churn

np.float64(1.2216593879412643)

In [118]:
churn_partner / global_churn

np.float64(0.7594724924338315)

Risk ratio > 1 are more likely to churn
with this we can see how each dataframe are important or less important but we can not do this for all columns so we can use groupby to find avg as example below

In [125]:
df_full_train.groupby('gender').churn.agg(['mean', 'count']) # agg() to get header form instead of series

mean  count
gender                 
female  0.276824   2796
male    0.263214   2838

In [ ]:
from IPython.display import display

In [128]:
# so we will do it for all catagorical values and diff and risk to each
for c in catagorical:
    df_group = df_full_train.groupby(c).churn.agg(['mean', 'count'])
    df_group['diff'] = df_group['mean'] - global_churn
    df_group['risk'] = df_group['mean'] / global_churn
    display(df_group)

mean  count      diff      risk
gender                                     
female  0.276824   2796  0.006856  1.025396
male    0.263214   2838 -0.006755  0.974980

mean  count      diff      risk
seniorcitizen                                     
0              0.242270   4722 -0.027698  0.897403
1              0.413377    912  0.143409  1.531208

mean  count      diff      risk
partner                                     
no       0.329809   2932  0.059841  1.221659
yes      0.205033   2702 -0.064935  0.759472

mean  count      diff      risk
dependents                                     
no          0.313760   3968  0.043792  1.162212
yes         0.165666   1666 -0.104302  0.613651

mean  count      diff      risk
phoneservice                                     
no            0.241316    547 -0.028652  0.893870
yes           0.273049   5087  0.003081  1.011412

mean  count      diff      risk
multiplelines                                        
no                0.257407   2700 -0.012561  0.953474
no_phone_service  0.241316    547 -0.028652  0.893870
yes               0.290742   2387  0.020773  1.076948

mean  count      diff      risk
internetservice                                     
dsl              0.192347   1934 -0.077621  0.712482
fiber_optic      0.425171   2479  0.155203  1.574895
no               0.077805   1221 -0.192163  0.288201

mean  count      diff      risk
onlinesecurity                                          
no                   0.420921   2801  0.150953  1.559152
no_internet_service  0.077805   1221 -0.192163  0.288201
yes                  0.153226   1612 -0.116742  0.567570

mean  count      diff      risk
onlinebackup                                            
no                   0.404323   2498  0.134355  1.497672
no_internet_service  0.077805   1221 -0.192163  0.288201
yes                  0.217232   1915 -0.052736  0.804660

mean  count      diff      risk
deviceprotection                                        
no                   0.395875   2473  0.125907  1.466379
no_internet_service  0.077805   1221 -0.192163  0.288201
yes                  0.230412   1940 -0.039556  0.853480

mean  count      diff      risk
techsupport                                             
no                   0.418914   2781  0.148946  1.551717
no_internet_service  0.077805   1221 -0.192163  0.288201
yes                  0.159926   1632 -0.110042  0.592390

mean  count      diff      risk
streamingtv                                             
no                   0.342832   2246  0.072864  1.269897
no_internet_service  0.077805   1221 -0.192163  0.288201
yes                  0.302723   2167  0.032755  1.121328

mean  count      diff      risk
streamingmovies                                         
no                   0.338906   2213  0.068938  1.255358
no_internet_service  0.077805   1221 -0.192163  0.288201
yes                  0.307273   2200  0.037305  1.138182

mean  count      diff      risk
contract                                           
month-to-month  0.431701   3104  0.161733  1.599082
one_year        0.120573   1186 -0.149395  0.446621
two_year        0.028274   1344 -0.241694  0.104730

mean  count      diff      risk
paperlessbilling                                     
no                0.172071   2313 -0.097897  0.637375
yes               0.338151   3321  0.068183  1.252560

mean  count      diff      risk
paymentmethod                                                 
bank_transfer_(automatic)  0.168171   1219 -0.101797  0.622928
credit_card_(automatic)    0.164339   1217 -0.105630  0.608733
electronic_check           0.455890   1893  0.185922  1.688682
mailed_check               0.193870   1305 -0.076098  0.718121

mean  count      diff      risk
churn                                 
0       0.0   4113 -0.269968  0.000000
1       1.0   1521  0.730032  3.704142

# 06 Feature importance: Mutual information
Mutual information - concept from information theory, it tells us how much we can learn about one variable if we know the value of another
- https://en.wikipedia.org/wiki/Mutual_information